src/picongpu/include/plugins/particleDetector/ParticleDetector.hpp

```diff
@@ -0,0 +1,537 @@
+
+        HBufHistogram hBufLeftParsHistogram(this->dBufLeftParsHistogram->size());
+
+        int rank;
+        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+        if(rank == 0)
+        {
+            splash::SerialDataCollector hdf5DataFile(1);
+            splash::DataCollector::FileCreationAttr fAttr;
+
+            splash::DataCollector::initFileCreationAttr(fAttr);
+            fAttr.fileAccType = splash::DataCollector::FAT_READ;
+
+            std::stringstream filename;
+            filename << restartDirectory << "/" << prefix << restartStep;
+
+            hdf5DataFile.open(filename.str().c_str(), fAttr);
+
+            splash::Dimensions dimensions;
+
+            hdf5DataFile.read(restartStep,
+                              "detectorLeftPars",
+                              dimensions,
+                              &(*hBufLeftParsHistogram.origin()));
+
+            hdf5DataFile.close();
+
+            /* rank 0 divides and distributes the detector to all ranks in equal parts */
+            int numRanks;
+            MPI_Comm_size(MPI_COMM_WORLD, &numRanks);
+            using namespace lambda;
+            PMacc::algorithm::host::Foreach()(hBufLeftParsHistogram.zone(),
+                                              hBufLeftParsHistogram.origin(),
+                                              _1 = _1 / float_X(numRanks));
+        }
+
+        MPI_Bcast(&(*hBufLeftParsHistogram.origin()),
+                  hBufLeftParsHistogram.size().productOfComponents() * sizeof(float_X),
+                  MPI_CHAR,
+                  0,
+                  MPI_COMM_WORLD);
+
+        *this->dBufLeftParsHistogram = hBufLeftParsHistogram;
+    }
+
+
+    void checkpoint(uint32_t currentStep, const std::string checkpointDirectory)
+    {
+        if(this->notifyPeriod == 0)
+            return;
+
+        HBufHistogram hBufLeftParsHistogram(this->dBufLeftParsHistogram->size());
+        HBufHistogram hBufTotal(hBufLeftParsHistogram.size());
+
+        hBufLeftParsHistogram = *this->dBufLeftParsHistogram;
+
+        /* mpi reduce */
+        using namespace lambda;
+        (*this->allGPU_reduce)(hBufTotal, hBufLeftParsHistogram, _1 + _2);
+        if(!this->allGPU_reduce->root())
+            return;
+
+        splash::SerialDataCollector hdf5DataFile(1);
+        splash::DataCollector::FileCreationAttr fAttr;
+
+        splash::DataCollector::initFileCreationAttr(fAttr);
+
+        std::stringstream filename;
+        filename << checkpointDirectory << "/" << prefix << currentStep;
+
+        hdf5DataFile.open(filename.str().c_str(), fAttr);
+
+        typename PICToSplash<float_X>::type SplashTypeX;
+
+        splash::Dimensions bufferSize(hBufTotal.size().x(),
+                                      hBufTotal.size().y(),
+                                      hBufTotal.size().z());
+
+        hdf5DataFile.write(currentStep,
+                           SplashTypeX,
+                           this->numBinsEnergy == 1 ? DIM2 : DIM3,
```

Comment:
I understand what you are doing here but it is in-place and obfuscated.

write this in a const variable above, add a comment string and explain that numBinsEnergy with a size larger then 1 means you will *actually do an energy binning* on not just a spatially/radially resolved energy-sum